internals/overlord/certstate/manager.go

```diff
@@ -0,0 +1,429 @@
+// whenever a cert change event took place (for example, when an
+// expired certificate gets replaced).
+func (c *CertManager) AddChangeListener(f CertChangedFunc) {
+	c.changeListeners = append(c.changeListeners, f)
+}
+
+func (c *CertManager) callChangeListeners(cert *x509.Certificate) {
+	if cert == nil {
+		// Avoids if statement on every deferred call to this method (we
+		// shouldn't call listeners when the operation fails).
+		return
+	}
+	for _, f := range c.changeListeners {
+		f(cert)
+	}
+}
+
+// newX509ECP256Keypair generates an elliptic P256 private key and an
+// X509 self-signed public certificate (signed with the private key). The x509
+// certificate includes both the public key and the signature, which allows
+// the certificate signature to be verified with the public key inside the
+// certificate. This allows a client side copy of the certificate to be used
+// as certificate authority for future TLS sessions with the server.
+func (c *CertManager) newX509ECP256Keypair() (*x509KeyPair, error) {
+	// Get the highest available order
+	order := 1
+	certCount := len(c.x509KeyPairs)
+	if certCount > 0 {
+		// Find the next available order
+		order = c.x509KeyPairs[certCount-1].Order + 1
+		if order > 999 {
+			return nil, fmt.Errorf("cannot process order number %v (valid range is 001-999)", order)
+		}
+	}
+
+	// Valid date range.
+	notBefore := time.Now()
+	notAfter := notBefore.Add(365 * 24 * time.Hour)
+	keypair, err := generateX509ECP256Keypair(notBefore, notAfter)
+	if err != nil {
+		return nil, err
+	}
+	keypair.Order = order
+	return keypair, nil
+}
+
+func (c *CertManager) persistX509Keypair(keypair *x509KeyPair) error {
+	return writeX509Keypair(keypair, c.tlsDir)
+}
+
+// PEMCertificates returns the slice of valid X509 certificates
+// available at the point where the call is made. The result is
+// a list of PEM encoded ceritficates.
+func (c *CertManager) PEMCertificates() ([][]byte, error) {
+	c.x509KeyPairsLock.Lock()
+	defer c.x509KeyPairsLock.Unlock()
+
+	var pemCerts [][]byte
+	for _, keypair := range c.x509KeyPairs {
+		if !isCertExpired(keypair.Certificate) {
+			pem, err := x509CertToPEM(keypair.Certificate)
+			if err != nil {
+				return nil, err
+			}
+			pemCerts = append(pemCerts, pem)
+		}
+	}
+	return pemCerts, nil
+}
+
+// Ensure implements StateManager.Ensure.
+func (c *CertManager) Ensure() error {
+
+	// TODO: Implement expired certificate purging, so that expired files
+	// are removed, and the ordering recalculated so we do not run out of
+	// order room.
+	return nil
+}
+
+// loadX509KeyPairs loads all the persisted X509 keypairs, without
+// changing the state of the system (it does not create new keypairs).
+func (c *CertManager) loadX509KeyPairs() error {
+	// If the directory does not exist that simply means no keypairs
+	// are currently available, and that is OK.
+	_, err := os.Stat(c.tlsDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+	}
```

Comment:
nit:
```suggestion
	if _, err := os.Stat(c.tlsDir); err != nil && os.IsNotExist(err) {
		return nil
	}
```

I'm always inclined to inline function calls returning errors in the `if` clause, when the line does not get too long. However, you can do as you please. I think we should enforce this style automatically, or at least have a team-wide convention on this specifically (because right now the KernOS codebase is inconsistent in this, largely my fault)